kolibri/core/assets/src/views/KModal.vue

```diff
@@ -228,16 +228,26 @@
           this.scrollShadow = this.maxContentHeight < this.$refs.content.scrollHeight;
         }
       }, 50),
+      // Emitted when the cancel button is clicked or the esc key is pressed
       emitCancelEvent() {
         if (!this.cancelDisabled) {
-          // Emitted when the cancel button is clicked or the esc key is pressed
-          this.$emit('cancel');
+          // If no @cancel is set on KModal, have the parent emit the event
+          if (this.$listeners.cancel) {
+            this.$emit('cancel');
+          } else {
+            this.$parent.$emit('cancel');
+          }
         }
       },
+      // Emitted when the submit button or the enter key is pressed
       emitSubmitEvent() {
         if (!this.submitDisabled) {
-          // Emitted when the submit button or the enter key is pressed
-          this.$emit('submit');
+          // If no @submit is set on KModal, have the parent emit the event
+          if (this.$listeners.submit) {
+            this.$emit('submit');
+          } else {
+            this.$parent.$emit('submit');
```

Comment:
I believe it should be the responsibility of the wrapper component to define this behavior because having a child modify the behavior of its parent goes against the typical component hierarchy model.

I haven't tried this before, but I think [you can use `v-on="$listeners"`](https://vuejs.org/v2/api/#vm-listeners) like this:

```html
<!-- AssignmentDeleteModal.vue -->
<template>
  <KModal
    :title="modalTitle"
    :submitText="coachStrings.$tr('deleteAction')"
    :cancelText="coachStrings.$tr('cancelAction')"
    v-on="$listeners"
  >
    <p>{{ modalDescription }}</p>
    <p v-if="modalConfirmation">
      {{ modalConfirmation }}
    </p>
  </KModal>
</template>
```

A similar strategy can be used for creating custom wrappers around native HTML elements like `<input>`. 

ref: https://github.com/vuejs/vue/issues/5983

Reply:
Using `v-on=$listeners` seems to work, but is kind of odd since instead of the Wrapper's child (KModal) defining the event passthrough behavior it's now the parent defining the passthrough behavior, namely by getting the event names right:

```
<ModalUser>
    <!-- "submit" will go through, but not "cancel", since the typo -->
    <WrappedModal @submit="handleSubmit" @carncel="handleCancel" />
        <KModal v-on="$listeners />
```

Another thing I liked about the `$parent.$emit` hack is that you don't get duplicated `$emit` events in the devtools

Reply:
Okay, I can try using `v-on=$listeners` in this refactor, which gives us, in my opinion, three benefits

1. Introduces an interesting new syntax into the codebase that might useful for other things.
1. Still gets rid of some boilerplate
1. Still encourages the consistent use of the events "submit" and "cancel" for all Modal-type components.